llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3834,11 +3834,13 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
     // minnum/maxnum intrinsics.
     if (SIFPOp->hasNoNaNs() && SIFPOp->hasNoSignedZeros()) {
       Value *X, *Y;
-      if (match(&SI, m_OrdFMax(m_Value(X), m_Value(Y))))
+      if (match(&SI, m_OrdFMax(m_Value(X), m_Value(Y))) ||
+          match(&SI, m_UnordFMax(m_Value(X), m_Value(Y))))
```

Comment:
It's a bit clumsy to have to match the two. If we're going to have dedicated matchers for it, is it worth having an OrdOrUnord variant?

Reply:
Okay. I'll refactor this code.
Should I update other uses of `m_OrdFMax` and `m_UnordFMax` uses in this patch?

Reply:
It's not that bad, can do in a follow up if it's worth it. I haven't looked at the other users of these, maybe they are missing this case too?

Reply:
I found 2 more uses of `m_OrdFMax` and both of them are combining `m_OrdFMax` with `m_UnordFMax`. It looks like the InstCombine is the only place where pattern matching is incomplete.

https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/ValueTracking.cpp#L8244-L8263

https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/IVDescriptors.cpp#L695-L696 and 
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/IVDescriptors.cpp#L699-L700.

Based on the description of `m_OrdFMax` and `m_UnodFMax` they must always be applied together. So, in my opinion, it's worth adding `m_OrdOrUnordFMax` and probably remove existing matchers to avoid similar mistakes in the future.

Reply:
Added new matchers in https://github.com/llvm/llvm-project/pull/112088/commits/1713edf3bb089296013f6fd6acd088bc17dd1dc8.